### ISSUE TITLE
[TypeScript] Produced draft is read-only if original state is read-only

### ISSUE THREAD

**Issue by olavim** [mweststrate/immer#97]
The `draft` returned by the `produce` function's callback should make a mutable clone from the original state. Currently immutability is retained:

```ts
interface IState {
  readonly foo: boolean;
}

const state: IState = {
  foo: false
};

produce(state, draft => {
  draft.foo = true; // Error: Cannot assign to 'foo' because it is a constant or a read-only property.
});
```

**Comment by tilastokeskus** [comment]
I submitted a PR to fix this.

As a quick solution until this is fixed in a release:

```ts
// custom.d.ts
declare module 'immer' {
  export type MutableState<T extends {[x: string]: any}, K extends string> = {
    [P in K]: T[P];
  };

  export type Recipe<S> = (this: S, draftState: MutableState<S, keyof S>) => void;

  export default function<S = any>(
    currentState: S,
    recipe: Recipe<S>
  ): S;
}
```

**Comment by ogwh** [comment]
This is fixed by https://github.com/mweststrate/immer/pull/109

**Comment by mweststrate** [comment]
Released as 1.4.0. Have fun!
